File: 207-course-schedule/207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegree = new int[numCourses];
        HashMap<Integer, List<Integer>> adjList = new HashMap<>();
        
        for(int[] edge : prerequisites){
            indegree[edge[0]]+=1;
            List<Integer> neighbours = adjList.getOrDefault(edge[1], new ArrayList<Integer>());
            neighbours.add(edge[0]);
            adjList.put(edge[1], neighbours);
        }
        
        Queue<Integer> queue = new LinkedList<>();
        for(int i=0;i<numCourses;i++){
            if(indegree[i] == 0){
                queue.offer(i);
            }
        }
        
        int[] topSort = new int[numCourses];
        int i = 0;
        while(!queue.isEmpty()){
            int cur = queue.remove();
            topSort[i++] = cur;
            if (adjList.containsKey(cur)) {
                for(int node : adjList.get(cur)){
                    indegree[node]-=1;
                    if(indegree[node]==0){
                        queue.offer(node);
                    }
                }
            }
        }
        
        if(i==numCourses){
            return true;
        }
        return false;
    }
}

File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        def topSort(indegree, conn):
            
            queue = deque()
            for node in range(len(indegree)):
                if indegree[node]==0:
                    queue.append(node)
            
            topOrder = []
            while queue:
                cur = queue.popleft()
                topOrder.append(cur)
                for nei in conn[cur]:
                    indegree[nei] -= 1
                    if indegree[nei] == 0:
                        queue.append(nei)
                        
            return len(topOrder) == len(indegree)
        
        graph = defaultdict(list)
        indegree = [0]*numCourses
        for a, b in prerequisites:
            graph[a].append(b)
            indegree[b]+=1
        
        return topSort(indegree, graph)
        
                    